libuavcan/include/uavcan/util/avl_tree.hpp

```diff
@@ -0,0 +1,387 @@
+        } else {
+            if (node->equal_keys == UAVCAN_NULLPTR) {
+                if (node->left == UAVCAN_NULLPTR || node->right == UAVCAN_NULLPTR) {
+                    Node *temp = node->left ? node->left : node->right;
+
+                    if (temp == UAVCAN_NULLPTR) {
+                        temp = node;
+                        node = UAVCAN_NULLPTR;
+                    } else {
+                        *node = *temp;
+                    }
+
+                    len_--;
+                    deleteNode(temp);
+                } else {
+                    Node *minOfRight = node->right;
+
+                    while (minOfRight->left != UAVCAN_NULLPTR) {
+                        minOfRight = minOfRight->left;
+                    }
+
+                    node->data = minOfRight->data;
+                    node->right = removeNode(node->right, minOfRight->data);
+                }
+            } else {
+                Node* newHead = deleteFromList(node, data);
+                return newHead;
+            }
+        }
+
+        if (node == UAVCAN_NULLPTR) {
+            return node;
+        }
+
+        node->h = static_cast<int16_t>(maxOf(heightOf(node->left), heightOf(node->right)) + 1);
+
+        int16_t balance = balanceOf(node);
+
+        if (balance > 1 && balanceOf(node->left) >= 0) {
+            return rotateRight(node);
+        }
+
+        if (balance > 1 && balanceOf(node->left) < 0) {
+            node->left = rotateLeft(node->left);
+            return rotateRight(node);
+        }
+
+        if (balance < -1 && balanceOf(node->right) <= 0) {
+            return rotateLeft(node);
+        }
+
+        if (balance < -1 && balanceOf(node->right) > 0) {
+            node->right = rotateRight(node->right);
+            return rotateLeft(node);
+        }
+
+        return node;
+    }
+
+    static bool linkedListContains(Node* head, const T* data) {
+        Node* next = head;
+        while(next != UAVCAN_NULLPTR) {
+            if (next->data == data) { /* Memory address comparison */
+                return true;
+            }
+            next = next->equal_keys;
+        }
+        return false;
+    }
+
+public:
+    AvlTree(IPoolAllocator &allocator, std::size_t allocator_quota)
+        : root_(UAVCAN_NULLPTR)
+        , len_(0)
+        , allocator_(allocator, allocator_quota)
+    {}
+
+    virtual ~AvlTree() {
+        // Delete leafs first.
+        postOrderTraverseNodeCleanup(this->root_);
+    }
+
+    void removeEntry(T* data) {
+        root_ = removeNode(root_, data);
+    }
+
+    bool insert(T* data) {
+        Node* newNode = makeNode(data);
+        if (newNode == UAVCAN_NULLPTR) {
+            return false;
+        }
+
+        root_ = insertNode(root_, newNode);
+
+        return true;
+    }
+
+    size_t getSize() const {
+        return root_ == UAVCAN_NULLPTR ? 0 : len_;
+    }
+
+    void walkPostOrder(std::function<void(T*&)> forEach) {
```

Comment:
Per my other comment, is this actually a useful function for anything other than testing?